### PR TITLE
UIORGS-200 Migrate to react-final-form

### PR DESCRIPTION
**Pull request by AndreiShumski2**
## Purpose
https://issues.folio.org/browse/UIORGS-200

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [ ] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/contacts/EditContact/EditContact.js [hunk 1]
```diff
@@ -260,7 +253,7 @@ const EditContact = ({
                 >
                   <AddressInfo
                     dropdownVendorCategories={categories}
-                    dispatchChange={dispatchChange}
+                    change={change}
```
Comment (aliaksei-chumakou) [inline src/contacts/EditContact/EditContact.js:256]: actually, we don't need to pass this as props. it's can be used anywhere with hooks, `useForm`

**Review by aliaksei-chumakou — APPROVED** [review]

File: src/Organizations/OrganizationForm/OrganizationAccountsForm/OrganizationAccountsForm.js [hunk 2]
```diff
@@ -55,7 +53,7 @@ class OrganizationAccountsForm extends Component {
   };
 
   renderSubForm = (elem, index, fields) => {
-    const isEditMode = Boolean(fields.get(index).id);
+    const isEditMode = Boolean(this.props.initialAccounts?.[index]?.name);
```
Comment (aliaksei-chumakou) [inline src/Organizations/OrganizationForm/OrganizationAccountsForm/OrganizationAccountsForm.js:56]: interesting...

File: src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js [hunk 3]
```diff
@@ -52,17 +51,17 @@ const EmailAddresses = ({ defaultLanguage, dispatchChange, dropdownVendorCategor
...
+              validate={validateRequired}
               valueKey={valueKey}
-              onSelect={(item) => {
-                fields.remove(index);
-                fields.insert(index, item);
-              }}
+              // eslint-disable-next-line no-unused-vars
+              onSelect={({ isPrimary, ...restItem }) => fields.update(index, restItem)}
+              validateFields={[]}
+              withFinalForm
```
Comment (aliaksei-chumakou) [inline src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js:64]: do we need this prop?

File: src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js [hunk 4]
```diff
@@ -52,17 +51,17 @@ const EmailAddresses = ({ defaultLanguage, dispatchChange, dropdownVendorCategor
...
               required
-              validate={[validateRequired]}
+              validate={validateRequired}
               valueKey={valueKey}
-              onSelect={(item) => {
-                fields.remove(index);
-                fields.insert(index, item);
-              }}
+              // eslint-disable-next-line no-unused-vars
+              onSelect={({ isPrimary, ...restItem }) => fields.update(index, restItem)}
```
Comment (aliaksei-chumakou) [inline src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js:62]: interesting...

File: src/Organizations/OrganizationForm/OrganizationForm.js [hunk 5]
```diff
@@ -31,21 +31,16 @@ import {
...
   handleSubmit,
-  dispatch,
-  change,
   initialValues,
   paneTitle,
   cancelForm,
-  isVendorForm,
-  formDefaultLanguage,
+  values: formValues,
+  fetchOrgByCode,
```
Comment (aliaksei-chumakou) [inline src/Organizations/OrganizationForm/OrganizationForm.js:43]: interesting...

File: src/Organizations/handleSaveErrorResponse.js [hunk 6]
```diff
@@ -15,7 +14,6 @@ export const handleSaveErrorResponse = async (intl, showCallout, response) => {
 
   const errorCode = responseJson.errors?.[0]?.code || ERROR_CODE_GENERIC;
   const errorMsg = responseJson.errors?.[0]?.message;
-  const errorKey = responseJson.errors?.[0]?.parameters?.[0]?.key;
```
Comment (aliaksei-chumakou) [inline src/Organizations/handleSaveErrorResponse.js:18]: interesting....

File: src/common/components/ButtonIsPrimary/ButtonIsPrimary.js [hunk 7]
```diff
@@ -1,23 +1,29 @@
...
+      fields.forEach((fieldName, i) => {
+        if (
+          (i !== fieldIndex && fields.value[i].isPrimary)
+          || i === fieldIndex
+        ) {
+          change(`${fieldName}.isPrimary`, !fields.value[i].isPrimary);
+        }
+      });
+    },
+    [fields, fieldIndex, change, isPrimary],
```
Comment (aliaksei-chumakou) [inline src/common/components/ButtonIsPrimary/ButtonIsPrimary.js:25]: are you sure that `fields` isn't a new object every render?

File: src/common/components/FieldIsPrimary/FieldIsPrimary.js [hunk 8]
```diff
@@ -1,19 +1,18 @@
...
-      fields.forEach((fieldName, i) => {
-        if (i !== fieldIndex && fields.get(i).isPrimary === newValue) {
-          dispatchChange(`${fieldName}.isPrimary`, !newValue);
-        }
-      });
-    }
+function FieldIsPrimary({ fields, fieldIndex, fieldPrefix, labelId, vertical }) {
+  const { change } = useForm();
+  const changeIsPrimary = ({ target: { checked } }) => {
+    fields.forEach((fieldName, i) => (
```
Comment (aliaksei-chumakou) [inline src/common/components/FieldIsPrimary/FieldIsPrimary.js:11]: what's wrong with previous code? `if (checked) {` isn't required?

File: test/bigtest/tests/organizationCreate.e2e.js [hunk 9]
```diff
@@ -125,29 +127,6 @@ describe('Create organization', function () {
           expect(orgDetails.accountsSection.acqUnitsView).to.contain('Test');
         });
       });
-
-      describe('no submit action since code contains spaces', () => {
```
Comment (aliaksei-chumakou) [inline test/bigtest/tests/organizationCreate.e2e.js:129]: interesting...

File: test/bigtest/tests/organizationEdit.e2e.js [hunk 10]
```diff
@@ -123,15 +127,17 @@ describe('Organization edit', () => {
     it('update Vendor Button is disabled', function () {
       expect(orgEdit.updateVendorButton.isDisabled).to.be.false;
     });
+  });
 
-    describe('click update vendor button', function () {
```
Comment (aliaksei-chumakou) [inline test/bigtest/tests/organizationEdit.e2e.js:127]: how this happens that some old tests are deleted?

File: src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js (hunk 3 above)
Comment (AndreiShumski2) [reply src/Organizations/OrganizationForm/OrganizationContactInfoForm/ContactInfoFormGroup/EmailAddresses.js:64]: Yes, moved it to [`stripes-acq-components`](https://github.com/folio-org/stripes-acq-components/pull/369)

**Review by aliaksei-chumakou — APPROVED** [review]

**Pull request merged**